gh-113876: Enhance itertools combinations

adriancretu opened the pull request:
Optimized code path for the case of just the last value needing an update. Moved loops to the worst-case branch (last value reached maximum).

Benchmarks against current implementation (I used Cython to test out the raw C implementation, code at the end).
To rule out the next() overhead time related to creation of new tuples the iterations dismiss the returned tuple's object reference.
To rule out initialization overhead, the input sequence for the combination is pre-created as a list.

Now, I'm not sure why the current implementation starts to bottleneck as I increase the K range and becomes slower than the Python' itertools.combinations() - but in any case, I have not found any instance where the proposed changes are in any way slower than the current. Speed ups may vary depending on how well K and N are chosen. 

```shell



[*25] (60, 5) Time for itertools.combinations: 0.609477 s
[*25] (60, 5) Time for combinations_next_new: 0.581689 s
[*25] (60, 5) Time for consume_combinations_next_old: 0.626798 s
Passed, 5461512 elements

[*25] (50, 6) Time for itertools.combinations: 1.862611 s
[*25] (50, 6) Time for combinations_next_new: 1.703859 s
[*25] (50, 6) Time for consume_combinations_next_old: 1.921321 s
Passed, 15890700 elements

[*25] (34, 9) Time for itertools.combinations: 6.789133 s
[*25] (34, 9) Time for combinations_next_new: 6.219420 s
Passed, 52451256 elements

[*1] (70, 7) Time for itertools.combinations: 5.495806 s
[*1] (70, 7) Time for combinations_next_new: 5.158365 s
[*1] (70, 7) Time for consume_combinations_next_old: 5.736585 s
Passed, 1198774720 elements

[*25] (73, 5) Time for itertools.combinations: 1.651183 s
[*25] (73, 5) Time for combinations_next_new: 1.591440 s
[*25] (73, 5) Time for consume_combinations_next_old: 1.693157 s
Passed, 15020334 elements

[*25] (500, 3) Time for itertools.combinations: 2.354216 s
[*25] (500, 3) Time for combinations_next_new: 2.177287 s
[*25] (500, 3) Time for consume_combinations_next_old: 2.209204 s
Passed, 20708500 elements

[*25] (1000, 2) Time for itertools.combinations: 0.055919 s
[*25] (1000, 2) Time for combinations_next_new: 0.052389 s
[*25] (1000, 2) Time for consume_combinations_next_old: 0.052676 s
Passed, 499500 elements
```


```c
cdef consume_combinations_next_old(object r, int k) noexcept:
    cdef object co = itertools_combinations_impl(r, k);
    cdef PyObject* t;

    while True:
        t = combinations_next_old(co);
        if t == NULL:
            break
        Py_DECREF(<object>t)


cdef consume_combinations_next_new(object r, int k) noexcept:
    cdef object co = itertools_combinations_impl(r, k);
    cdef PyObject* t;

    while True:
        t = combinations_next_new(co)
        if t == NULL:
            break
        Py_DECREF(<object>t)


cdef consume_itertools_combinations(r, k) noexcept:
    deque(combinations(r, k), maxlen=0)


def benchmark_combinations(int n, int k, int repeat):
    r = list(range(n))

    t = timeit.timeit(stmt=lambda: consume_itertools_combinations(r, k), number=repeat)
    print(f"[*{repeat}] {n, k} Time for itertools.combinations: {t:.6f} s")

    t = timeit.timeit(stmt=lambda: consume_combinations_next_new(r, k), number=repeat)
    print(f"[*{repeat}] {n, k} Time for combinations_next_new: {t:.6f} s")

    t = timeit.timeit(stmt=lambda: consume_combinations_next_old(r, k), number=repeat)
    print(f"[*{repeat}] {n, k} Time for consume_combinations_next_old: {t:.6f} s")

    test_combinations_changes(r, k)


def test_combinations_changes(r, int k):
    cdef truth = combinations(r, k)
    cdef object co = itertools_combinations_impl(r, k);
    cdef int total = 0

    for combo in truth:
        candidate = <tuple> combinations_next_patch(co)
        if candidate != combo:
            raise Exception(f'Validation failed! {combo} != {candidate}')
        total += 1
        Py_DECREF(<object> candidate)

    print(f'Passed, {total} elements')
```



<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-113876 -->
* Issue: gh-113876
<!-- /gh-issue-number -->

adriancretu committed:
Update itertoolsmodule.combinations_next

Optimized code path for the case of just the last value needing an update. Moved loops to the worst-case branch (last value reached maximum).

adriancretu committed:
Fix bad comment.

Fix preprocessing error.

adriancretu committed:
Fix 0-length indices.

erlend-aasland commented on Modules/itertoolsmodule.c:
```diff
@@ -2666,6 +2666,14 @@ combinations_traverse(combinationsobject *co, visitproc visit, void *arg)
+    oldelem = PyTuple_GET_ITEM(result, i);      \
+    PyTuple_SET_ITEM(result, i, elem);          \
+    Py_DECREF(oldelem);
+
```
Please move this macro down to where it is used (for example, just above `combinations_next()`), and remember to `#undef` it after use.

erlend-aasland replied:
Also, it is generally better to write such macros using the C macro idiom `#define MACRO(...) do { /* body */ } while (0)`. See PEP-7 for more details.

adriancretu committed:
(lint) trim whitespace

adriancretu committed:
Add undefine

erlend-aasland commented on Modules/itertoolsmodule.c:
```diff
@@ -2714,33 +2723,48 @@ combinations_next(combinationsobject *co)
         assert(r == 0 || Py_REFCNT(result) == 1);
 
+        /* Seems like we allow 0-length combination ranges. */
+        if (0 == r) goto empty;
```
```suggestion
        if (0 == r) {
            goto empty;
        }
```

erlend-aasland commented on Modules/itertoolsmodule.c:
```diff
@@ -2714,33 +2723,48 @@ combinations_next(combinationsobject *co)
+        if (indices[i] == i + n - r) {
+            /* If we only have one index, at max value, we're done. */
+            if (0 == i)
+                goto empty;
```
```suggestion
            if (0 == i) {
                goto empty;
            }
```

erlend-aasland commented on Modules/itertoolsmodule.c:
```diff
@@ -2714,33 +2723,48 @@ combinations_next(combinationsobject *co)
+            for (;;) {
+                --i;
+                if (indices[i] < i + n - r)
+                    break;
```
```suggestion
                if (indices[i] < i + n - r) {
                    break;
                }
```

erlend-aasland commented on Modules/itertoolsmodule.c:
```diff
@@ -2714,33 +2723,48 @@ combinations_next(combinationsobject *co)
+                /* If i is zero, then the indices are all at
+                   their maximum value and we're done. */
+                if (0 == i)
+                    goto empty;
```
```suggestion
                if (0 == i) {
                    goto empty;
                }
```

erlend-aasland commented on Modules/itertoolsmodule.c:
```diff
@@ -2714,33 +2723,48 @@ combinations_next(combinationsobject *co)
+            for (;;) {
+                ITERTOOLS_SET_COMBINATION_VALUE(i);
+                if (i == r - 1)
+                    break;
```
```suggestion
                if (i == r - 1) {
                    break;
                }
```

adriancretu committed:
Apply suggestions from code review

Co-authored-by: Erlend E. Aasland <erlend.aasland@protonmail.com>

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

erlend-aasland commented on Misc/NEWS.d/next/Library/2024-01-11-14-06-56.gh-issue-113876.wUfzO6.rst:
```diff
@@ -0,0 +1 @@
+Improve ``itertools.combinations`` efficiency, making it faster when iterating over a large number of combinations.
```
```suggestion
Improve :py:func:`itertools.combinations` efficiency, making it faster when iterating over a large number of combinations.
```

erlend-aasland committed:
Update Misc/NEWS.d/next/Library/2024-01-11-14-06-56.gh-issue-113876.wUfzO6.rst

adriancretu committed:
Cache constant values

Cache combination's (n - r) between next() calls since it's a static value; cache n for combination_with_replacement for same reason.

adriancretu committed:
Lint fix

adriancretu committed:
Fix test

Adapt size test to structs having one new element.

adriancretu committed:
Merge branch 'main' into enhance_combinations

tim-one commented on Modules/itertoolsmodule.c:
```diff
@@ -2712,33 +2723,55 @@ combinations_next(combinationsobject *co)
          */
         assert(r == 0 || Py_REFCNT(result) == 1);
 
+        /* Seems like we allow 0-length combination ranges. */
```
Since `comb(n, 0) == 1`, yes, whenever `r == 0` we produce one result (an empty tuple). So don't say "seems like" 😉 I think it would be better to move `if (0 == r) goto empty;` to the top of this `else` block, and get rid of that special case at once. We already returned an empty tuple the first time this function was called. The only thing to do on the second call in the `r == 0` case is `goto empty`. All the code before this line in the `else` block is useless churn when `r == 0`. Very rarely executed, but useless all the same,  so it would be clearer to skip it. Then the `assert` can lose the `r == 0` clause too.

tim-one commented on Modules/itertoolsmodule.c:
```diff
@@ -2712,33 +2723,55 @@ combinations_next(combinationsobject *co)
+                indices[i] = indices[i - 1] + 1;
+            }
+        }
+        else {
```
```suggestion
        else {
        /* This is the usual case, and keeping it separate from the
            search loops above typically speeds the function overall
            by at least a few percent. */
```

tim-one commented on Modules/itertoolsmodule.c:
```diff
@@ -2570,6 +2570,7 @@ typedef struct {
     PyObject *pool;         /* input converted to a tuple */
     Py_ssize_t *indices;    /* one index per result element */
     PyObject *result;       /* most recently returned result tuple */
+    Py_ssize_t n_r_delta;   /* maximum possible value for first index (n - r) */
```
Does adding the `n_r_delta` member buy something of value? Computing `n-r` when needed  is very cheap, should overlap with memory reads anyway, and should be easy for an optimizing compiler to recognize as an invariant subexpression.

adriancretu replied:
It is a precomputed value used in a few places which can loop (including next() itself), so whether re-computation is actually optimized or not by the compiler is an open question, but I doubt it.

When called millions of times in a row, which is what I understand what combinations can end up to, the "cheap" re-computation at every step can overtake the small penalty of an extra combinations state variable.

tim-one replied:
It won't be reused _across_ calls, no. The only way to know is to time it. If it actually does help, then instead of adding a new member to the struct, I'm sure it would do just as well to add, e.g.,

```c
    const Py_ssize_t n_r_delta = n - r;
```

near the top of the function. That tells the compiler directly that it's a function-invariant value, even in a debug build, and without putting additional pressure on caches.

tim-one replied:
And noting that computing `n-r` at all is something of a red herring. Regardless of `r`, the largest value at the highest index in the tuple is `n-1`. So instead of bothering with `n-r`, a variable (say `hi`) could be initialized to `n-1`, and decremented by 1 each time through the search loop. I believe that would also be more obvious.

adriancretu replied:
The benchmarks and this PR were done in January, so whether your statements are correct or not can be proven by running new benchmarks. What I did then was to move any "constant" values off the processing logic. This was my first Python PR so if you feel that using a "const" has the identical effect as the intent of this change, then by all means pursue it. I just thought the co object was the best place to put such an invariable value. I also found that small changes to how many local variables a low-level loop uses can affect the resulting assembly code (and hence the speed) in a lot of ways. YMMV.

tim-one replied:
I apologize on behalf of the core devs that this sat unaddressed for so long. When Raymond added me, I leaped right on it.

But, for good or ill,, the burden remains on you to "sell" it. "YMMV" is worse than you may think :wink: - for speedups so small, they vary across platforms (including HW, compiler, compiler release, and compiler options) too. 

BTW, under Microsoft's Visual Studio 2022 in a release build, I got similar speedups _just_ by introducing a `const` variable for `n-r` in the current implementation, and replacing the single instance of `n-r` with it. So that specific platform is doing a poor job at optimizing `n-r`.

It's a minor speedup at best, so I'm happy to put _some_ effort into moving this along, but not a lot of effort. If you're worn out from putting effort into it too, that's understandable, but I'll have to close the PR. If you want to pursue this, you'll need to address comments made.
